File: notify_server.js
#!/usr/bin/env node
/**
 * Tars Notify Server - Desktop notifications for Clawdbot
 * Plays sounds when tasks complete so you don't have to wait staring at the screen
 * 
 * Usage: node notify_server.js
 * Server runs on http://localhost:8765
 */

const http = require('http');
const fs = require('fs');
const path = require('path');
const { exec } = require('child_process');
const os = require('os');

const PORT = 8765;
const SOUNDS_DIR = path.join(__dirname, 'sounds');

// Ensure sounds directory exists
if (!fs.existsSync(SOUNDS_DIR)) {
    fs.mkdirSync(SOUNDS_DIR, { recursive: true });
}

// Check platform
const platform = os.platform();
let audioBackend = null;

// Detect available audio backends
function detectAudioBackend() {
    // Windows
    if (platform === 'win32') {
        audioBackend = 'powershell';
        return;
    }
    
    // macOS
    if (platform === 'darwin') {
        exec('which afplay', (err) => {
            if (!err) audioBackend = 'afplay';
        });
        return;
    }
    
    // Linux - check various
    const checkCmd = 'which aplay paplay play || echo "none"';
    exec(checkCmd, (err, stdout) => {
        if (stdout.includes('aplay')) audioBackend = 'aplay';
        else if (stdout.includes('paplay')) audioBackend = 'paplay';
        else if (stdout.includes('play')) audioBackend = 'sox';
    });
}

// Generate a simple WAV file
function generateWav(filename, freq = 440, duration = 0.3, pattern = 'beep') {
    const filepath = path.join(SOUNDS_DIR, filename);
    if (fs.existsSync(filepath)) return filepath;
    
    const sampleRate = 44100;
    const numSamples = Math.floor(duration * sampleRate);
    const dataSize = numSamples * 4; // 2 channels * 2 bytes
    const fileSize = 44 + dataSize;
    
    const buffer = Buffer.alloc(fileSize);
    let offset = 0;
    
    // WAV Header
    buffer.write('RIFF', offset); offset += 4;
    buffer.writeUInt32LE(fileSize - 8, offset); offset += 4;
    buffer.write('WAVE', offset); offset += 4;
    buffer.write('fmt ', offset); offset += 4;
    buffer.writeUInt32LE(16, offset); offset += 4; // Subchunk1Size
    buffer.writeUInt16LE(1, offset); offset += 2; // AudioFormat (PCM)
    buffer.writeUInt16LE(2, offset); offset += 2; // NumChannels (stereo)
    buffer.writeUInt32LE(sampleRate, offset); offset += 4;
    buffer.writeUInt32LE(sampleRate * 4, offset); offset += 4; // ByteRate
    buffer.writeUInt16LE(4, offset); offset += 2; // BlockAlign
    buffer.writeUInt16LE(16, offset); offset += 2; // BitsPerSample
    buffer.write('data', offset); offset += 4;
    buffer.writeUInt32LE(dataSize, offset); offset += 4;
    
    // Generate samples
    for (let i = 0; i < numSamples; i++) {
        const t = i / sampleRate;
        let value = 0;
        
        if (pattern === 'beep') {
            const envelope = i < numSamples * 0.1 ? 1 : 0.5;
            value = Math.floor(32767 * 0.3 * envelope * Math.sin(2 * Math.PI * freq * t));
        } else if (pattern === 'success') {
            const f = t < duration / 2 ? freq : freq * 1.5;
            const envelope = 1 - t / duration;
            value = Math.floor(32767 * 0.3 * envelope * Math.sin(2 * Math.PI * f * t));
        } else if (pattern === 'error') {
            const f = freq * (1 - t / duration * 0.5);
            value = Math.floor(32767 * 0.3 * Math.sin(2 * Math.PI * f * t));
        } else if (pattern === 'ping') {
            const envelope = Math.max(0, 1 - t / duration * 3);
            value = Math.floor(32767 * 0.2 * envelope * Math.sin(2 * Math.PI * freq * t));
        } else if (pattern === 'complete') {
            const segment = Math.floor(t / (duration / 3));
            const f = freq * (1 + segment * 0.5);
            const envelope = 1 - t / duration;
            value = Math.floor(32767 * 0.25 * envelope * Math.sin(2 * Math.PI * f * t));
        }
        
        // Stereo (left and right)
        buffer.writeInt16LE(value, offset); offset += 2;
        buffer.writeInt16LE(value, offset); offset += 2;
    }
    
    fs.writeFileSync(filepath, buffer);
    return filepath;
}

// Generate all sounds
function initSounds() {
    console.log('🔊 Generating sounds...');
    generateWav('success.wav', 880, 0.4, 'success');
    generateWav('error.wav', 440, 0.5, 'error');
    generateWav('ping.wav', 1200, 0.15, 'ping');
    generateWav('complete.wav', 660, 0.6, 'complete');
    console.log('✅ Sounds ready');
}

// Play sound
function playSound(soundName = 'success') {
    const filepath = path.join(SOUNDS_DIR, `${soundName}.wav`);
    if (!fs.existsSync(filepath)) {
        console.log(`⚠️ Sound not found: ${soundName}`);
        return;
    }
    
    let cmd = null;
    
    if (platform === 'win32') {
        // Use PowerShell to play WAV
        cmd = `powershell -c "(New-Object Media.SoundPlayer '${filepath}').PlaySync()"`;
    } else if (platform === 'darwin') {
        cmd = `afplay "${filepath}"`;
    } else {
        // Linux - try different backends
        if (audioBackend === 'aplay') {
            cmd = `aplay "${filepath}" 2>/dev/null`;
        } else if (audioBackend === 'paplay') {
            cmd = `paplay "${filepath}" 2>/dev/null`;
        } else if (audioBackend === 'sox') {
            cmd = `play "${filepath}" 2>/dev/null`;
        }
    }
    
    if (cmd) {
        exec(cmd, { timeout: 5000 }, (err) => {
            if (err) console.log(`🔇 Audio error: ${err.message}`);
        });
    } else {
        console.log(`🔇 Would play: ${soundName}`);
        // Terminal bell fallback
        process.stdout.write('\x07');
    }
}

// HTTP Server
const server = http.createServer((req, res) => {
    res.setHeader('Content-Type', 'application/json');
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Access-Control-Allow-Methods', 'GET, POST, OPTIONS');
    res.setHeader('Access-Control-Allow-Headers', 'Content-Type');
    
    if (req.method === 'OPTIONS') {
        res.writeHead(200);
        res.end();
        return;
    }
    
    const url = req.url;
    const timestamp = new Date().toISOString();
    
    // Status endpoint
    if (url === '/' || url === '/status') {
        res.writeHead(200);
        res.end(JSON.stringify({
            status: 'running',
            audio_backend: audioBackend || 'default',
            platform: platform,
            sounds: ['success', 'error', 'ping', 'complete'],
            port: PORT,
            timestamp
        }, null, 2));
        return;
    }
    
    // Notify endpoint
    if (url === '/notify' && req.method === 'POST') {
        let body = '';
        req.on('data', chunk => body += chunk);
        req.on('end', () => {
            try {
                const data = JSON.parse(body);
                const message = data.message || 'Notification';
                const sound = data.sound || 'success';
                const time = new Date().toLocaleTimeString();
                
                console.log(`🔔 [${time}] ${message}`);
                playSound(sound);
                
                res.writeHead(200);
                res.end(JSON.stringify({
                    ok: true,
                    message,
                    sound,
                    timestamp
                }));
            } catch (e) {
                res.writeHead(400);
                res.end(JSON.stringify({ error: 'Invalid JSON' }));
            }
        });
        return;
    }
    
    // Test sound endpoint
    if (url.startsWith('/test/')) {
        const soundName = url.split('/')[2];
        playSound(soundName);
        res.writeHead(200);
        res.end(JSON.stringify({ ok: true, sound: soundName }));
        return;
    }
    
    // Shutdown endpoint
    if (url === '/shutdown' && req.method === 'POST') {
        res.writeHead(200);
        res.end(JSON.stringify({ ok: true, message: 'Shutting down...' }));
        setTimeout(() => process.exit(0), 100);
        return;
    }
    
    // 404
    res.writeHead(404);
    res.end(JSON.stringify({ error: 'Not found' }));
});

// Start server
detectAudioBackend();
initSounds();

server.listen(PORT, '0.0.0.0', () => {
    console.log('='.repeat(50));
    console.log('🤖 Tars Notify Server');
    console.log('='.repeat(50));
    console.log(`Desktop notifications for Clawdbot`);
    console.log(`Server: http://localhost:${PORT}`);
    console.log('-'.repeat(50));
    console.log('Ready! Trigger with:');
    console.log(`  curl -X POST http://localhost:${PORT}/notify \\\n    -H "Content-Type: application/json" \\\n    -d '{"message": "Done!", "sound": "success"}'`);
    console.log('-'.repeat(50));
    
    // Play startup sound
    setTimeout(() => playSound('ping'), 500);
});

// Handle graceful shutdown
process.on('SIGINT', () => {
    console.log('\n🛑 Shutting down...');
    server.close(() => process.exit(0));
});

File: notify_server.py
#!/usr/bin/env python3
"""
Tars Notify Server - Desktop notifications for Clawdbot
Plays sounds when tasks complete so you don't have to wait staring at the screen.
"""

import os
import sys
import json
import wave
import struct
import threading
import subprocess
from datetime import datetime
from flask import Flask, request, jsonify

app = Flask(__name__)
PORT = 8765
SOUNDS_DIR = os.path.join(os.path.dirname(__file__), 'sounds')

# Ensure sounds directory exists
os.makedirs(SOUNDS_DIR, exist_ok=True)

# Try different audio backends
AUDIO_BACKEND = None

def init_audio():
    """Initialize audio backend"""
    global AUDIO_BACKEND
    
    # Try pygame first (cross-platform)
    try:
        import pygame
        pygame.mixer.init(frequency=44100, size=-16, channels=2, buffer=512)
        AUDIO_BACKEND = 'pygame'
        print("✅ Audio backend: pygame")
        return True
    except Exception as e:
        print(f"⚠️ pygame failed: {e}")
    
    # Try playsound
    try:
        from playsound import playsound
        AUDIO_BACKEND = 'playsound'
        print("✅ Audio backend: playsound")
        return True
    except Exception as e:
        print(f"⚠️ playsound failed: {e}")
    
    # Windows-specific
    if sys.platform == 'win32':
        try:
            import winsound
            AUDIO_BACKEND = 'winsound'
            print("✅ Audio backend: winsound")
            return True
        except Exception as e:
            print(f"⚠️ winsound failed: {e}")
    
    # macOS-specific
    if sys.platform == 'darwin':
        AUDIO_BACKEND = 'afplay'
        print("✅ Audio backend: afplay (macOS)")
        return True
    
    # Linux - try various
    for cmd in ['aplay', 'paplay', 'ogg123']:
        if subprocess.run(['which', cmd], capture_output=True).returncode == 0:
            AUDIO_BACKEND = cmd
            print(f"✅ Audio backend: {cmd}")
            return True
    
    print("⚠️ No audio backend available - notifications will be silent")
    return False

def generate_sound(filename, freq=440, duration=0.3, pattern='beep'):
    """Generate a simple WAV sound file"""
    filepath = os.path.join(SOUNDS_DIR, filename)
    if os.path.exists(filepath):
        return filepath
    
    sample_rate = 44100
    num_samples = int(duration * sample_rate)
    
    with wave.open(filepath, 'w') as wav:
        wav.setnchannels(2)
        wav.setsampwidth(2)
        wav.setframerate(sample_rate)
        
        for i in range(num_samples):
            t = i / sample_rate
            
            if pattern == 'beep':
                # Simple beep
                value = int(32767 * 0.3 * (1 if i < num_samples * 0.1 else 0.5) * 
                           (1 if i % int(sample_rate/freq) < int(sample_rate/freq)/2 else -1))
            elif pattern == 'success':
                # Ascending two-tone
                f = freq if t < duration/2 else freq * 1.5
                value = int(32767 * 0.3 * (1 - t/duration) * 
                           (1 if i % int(sample_rate/f) < int(sample_rate/f)/2 else -1))
            elif pattern == 'error':
                # Descending tone
                f = freq * (1 - t/duration * 0.5)
                value = int(32767 * 0.3 * 
                           (1 if i % int(sample_rate/f) < int(sample_rate/f)/2 else -1))
            elif pattern == 'ping':
                # Short high ping
                envelope = max(0, 1 - t/duration * 3)
                value = int(32767 * 0.2 * envelope * 
                           (1 if i % int(sample_rate/freq) < int(sample_rate/freq)/2 else -1))
            elif pattern == 'complete':
                # Three ascending tones
                segment = int(t / (duration / 3))
                f = freq * (1 + segment * 0.5)
                value = int(32767 * 0.25 * (1 - t/duration) * 
                           (1 if i % int(sample_rate/f) < int(sample_rate/f)/2 else -1))
            else:
                value = 0
            
            # Stereo
            wav.writeframes(struct.pack('<hh', value, value))
    
    return filepath

def init_sounds():
    """Generate default sound files"""
    print("🔊 Generating sounds...")
    generate_sound('success.wav', 880, 0.4, 'success')
    generate_sound('error.wav', 440, 0.5, 'error')
    generate_sound('ping.wav', 1200, 0.15, 'ping')
    generate_sound('complete.wav', 660, 0.6, 'complete')
    print("✅ Sounds ready")

def play_sound_file(filepath):
    """Play a sound file using available backend"""
    if not AUDIO_BACKEND:
        print(f"🔇 Would play: {filepath}")
        return
    
    try:
        if AUDIO_BACKEND == 'pygame':
            import pygame
            sound = pygame.mixer.Sound(filepath)
            sound.play()
        elif AUDIO_BACKEND == 'playsound':
            from playsound import playsound
            playsound(filepath, block=False)
        elif AUDIO_BACKEND == 'winsound':
            import winsound
            winsound.PlaySound(filepath, winsound.SND_FILENAME | winsound.SND_ASYNC)
        elif AUDIO_BACKEND == 'afplay':
            subprocess.Popen(['afplay', filepath])
        elif AUDIO_BACKEND in ['aplay', 'paplay', 'ogg123']:
            subprocess.Popen([AUDIO_BACKEND, filepath], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
    except Exception as e:
        print(f"🔇 Audio error: {e}")

def play_sound(sound_name='success'):
    """Play a named sound"""
    filepath = os.path.join(SOUNDS_DIR, f'{sound_name}.wav')
    if os.path.exists(filepath):
        play_sound_file(filepath)
    else:
        print(f"⚠️ Sound not found: {sound_name}")
        # Fallback to system beep
        if sys.platform == 'win32':
            import winsound
            winsound.MessageBeep()
        else:
            print('\a')  # Terminal bell

# Flask routes
@app.route('/')
def index():
    """Status page"""
    return jsonify({
        'status': 'running',
        'audio_backend': AUDIO_BACKEND or 'none',
        'sounds': ['success', 'error', 'ping', 'complete'],
        'port': PORT,
        'timestamp': datetime.now().isoformat()
    })

@app.route('/status')
def status():
    """Get server status"""
    return jsonify({
        'status': 'running',
        'audio_backend': AUDIO_BACKEND or 'none',
        'timestamp': datetime.now().isoformat()
    })

@app.route('/notify', methods=['POST'])
def notify():
    """Trigger a notification sound"""
    data = request.get_json() or {}
    message = data.get('message', 'Notification')
    sound_name = data.get('sound', 'success')
    
    print(f"🔔 [{datetime.now().strftime('%H:%M:%S')}] {message}")
    
    # Play sound in background thread
    thread = threading.Thread(target=play_sound, args=(sound_name,))
    thread.daemon = True
    thread.start()
    
    return jsonify({
        'ok': True,
        'message': message,
        'sound': sound_name,
        'timestamp': datetime.now().isoformat()
    })

@app.route('/test/<sound_name>')
def test_sound(sound_name):
    """Test a specific sound"""
    play_sound(sound_name)
    return jsonify({'ok': True, 'sound': sound_name})

@app.route('/shutdown', methods=['POST'])
def shutdown():
    """Shutdown the server"""
    func = request.environ.get('werkzeug.server.shutdown')
    if func:
        func()
    return jsonify({'ok': True, 'message': 'Shutting down...'})

def main():
    """Main entry point"""
    print("=" * 50)
    print("🤖 Tars Notify Server")
    print("=" * 50)
    print(f"Desktop notifications for Clawdbot")
    print(f"Server: http://localhost:{PORT}")
    print("-" * 50)
    
    init_audio()
    init_sounds()
    
    print("-" * 50)
    print("Ready! Trigger with:")
    print(f'  curl -X POST http://localhost:{PORT}/notify -H "Content-Type: application/json" -d \'{"message": "Done!", "sound": "success"}\'')
    print("-" * 50)
    
    # Open browser to status page
    try:
        import webbrowser
        webbrowser.open(f'http://localhost:{PORT}')
    except:
        pass
    
    # Run server
    app.run(host='0.0.0.0', port=PORT, debug=False)

if __name__ == '__main__':
    main()
